Add EIP-6051: Private Key Encapsulation

weiji-cryptonatty opened the pull request:
When opening a pull request to submit a new EIP, please use the suggested template: https://github.com/ethereum/EIPs/blob/master/eip-template.md

We have a GitHub bot that automatically merges some PRs. It will merge yours immediately if certain criteria are met:

 - The PR edits only existing draft PRs.
 - The build passes.
 - Your GitHub username or email address is listed in the 'author' header of all affected PRs, inside <triangular brackets>.
 - If matching on email address, the email address is the one publicly listed on your GitHub profile.

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,464 @@
+discussions-to: https://ethereum-magicians.org/t/private-key-encapsulation-to-move-around-securely-without-entering-seed/11604
+status: Draft
+type: Standards Track
+category: ERC
```
This seems an `Interface` category

weiji-cryptonatty replied:
I double checked eip-1 for `Interface` use. I don't think this should be an `Interface` category. The requests are optional for wallet applications. It does _not_ require _all_ wallet applications to implement. Also, EIP-5630 uses `ERC` category as well, while it also provides eth_ api additions like `eth_getEncryptionPublicKey ` and `eth_decrypt `.

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,464 @@
+
+## Abstract
+
+This EIP proposes a mechanism to encapsulate a private key so that it could be securely relocated to another application without providing the seed. This EIP combines `ECIES` and optional signature verification under various choices to ensure that the private key is encapsulated to a known or trusted party.
```
[Editorial] Suggest spell out `ECIES` what it's abbreviated for and provide relevant brief explanation

weiji-cryptonatty replied:
added

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,464 @@
+4.  Share secret  `ss := [s]R = [r]S` according to ECDH. Note that for secp256k1 this EIP follows RFC5903 and uses compact representation, which means to use *only* the `x` coordinate as the shared secret. For Curve25519 this EIP follows RFC7748.
+5.  `oob`, out of band data, optional. This could be digits or an alpha-numeric string entered by user.
+6.  Let  `derivedKey  :=  HKDF(hash=SHA256, ikm=ss, info=oob,  salt,  length)`. HKDF is defined in RFC5869. The  `length`  should be determined by  `skey`  and  `IV`  requirements such that the symmetric key  `skey  = derivedKey[0:keySize]`, and  `IV  = derivedKey[keySize:length]`.  `keySize`  denotes the key size of underlying symmetric algorithm, for example, 16 (bytes) for AES-128, and 32 (byt ...
+7.  `cipher  := authenticated_cncryption(symAlg,  skey,  IV, data=sk)`. The symmetric cipher algorithm `symAlg` and authentication scheme are decided by the version parameter. No additional authentication data `aad` is used.
```
```suggestion
7.  `cipher  := authenticated_encryption(symAlg,  skey,  IV, data=sk)`. The symmetric cipher algorithm `symAlg` and authentication scheme are decided by the version parameter. No additional authentication data `aad` is used.
```

xinbenlv commented on EIPS/eip-6051.md:
````diff
@@ -0,0 +1,464 @@
+	params: [version, signerPubKey],
+})
+```
+
````
Suggest putting in the response format here

weiji-cryptonatty replied:
Added. The `Requests` here are calls to window.ethereum object and then gets a return value as in a javascript function call:

```javascript
window.ethereum.request({
	method: 'eth_generateEphemeralKeyPair',
	params: [version, signerPubKey],
})
// expected return value: R
```

weiji-cryptonatty replied:
Also the test vectors shall clarify further.

xinbenlv commented on EIPS/eip-6051.md:
````diff
@@ -0,0 +1,464 @@
+```javascript
+request({
+	method: 'eth_generateEphemeralKeyPair',
+	params: [version, signerPubKey],
````
Specify the length of version and signerPubKey here or mention it will be specified below.

weiji-cryptonatty replied:
moved "Encoding of data and message" prior to "Requests".

weiji-cryptonatty commented:
Hi @xinbenlv I had pushed some more commits to address all issues except the `ERC` vs `Interface`. Could you please review again? Thanks a lot.

xinbenlv commented:
Glad to hear! 

I will continue to review the remaining 2/3 of the EIPs.

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+In both cases, it is safer not to provide the seed which controls the whole wallet and might contains many addresses in multiple chains.
+
+This EIP aims to enable such use cases.
```
So far Motivation is good enough for merge as a `Draft`.

Down the road, I think this EIP could still benefit a more thorough and elaborate demonstrated of use cases. Such as the scenarios when a cold wallet exports some private keys and to which third party. This helps reader understand the security model this EIP tries to address.

xinbenlv commented:
I am going to mainly focus on the editorial review for now, so as we could unblock it for merging as `draft`, but I do have a few technical questions / feedback when conducting my technical peer review, and they will come later

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+The return data is a byte array which contains first the ephemeral sender public key `S` (compressed if applicable), then `cipher` including any authentication tag, that is, `S || cipher`.
+
+#### R3. Request for Recipient to unwrap and intake the private key
```
For R3, in your typical use case, who is the "requester" and who is the "responder", and is the communication channel between them considered "secured?

weiji-cryptonatty replied:
You may think of it as a function call. All requests here (R1, R2, R3) are supposed to be called within a web page:

```javascript
window.ethereum.request){
...})
```

Wallets like MetaMask, Tally Ho  or any WalletConnect apps could be called up to process the request (I am not pretty sure about WalletConnect). A hardware/cold wallet may be working behind, connected in a secure manner.

weiji-cryptonatty replied:
Regarding the security or threat model, we need to prevent a malicious party from intercepting or tampering with the Recipient public key `R`, therefore the Sender application may reject the request R2 if no signature is provided, or not provided as expected (`signerPubKey` is not signed by `trustedPubKey`, just for example). But signature is not mandatory. And the intention is to provide implementation flexibility. For example, a messaging application as Recipient may not have a proper signature, in such cases, it is expected that the private key being encapsulated is dedicated for messaging purposes.

It is possible that Sender application (under attacker's control) provides a 'fake' or 'wrong' private key in R2, but then in R3 user will find out by checking out the returned address of the intake key. A private key installed by attacker may be used to decrypt confidential information or steal funds. This is unlikely, though, as if Sender application is already under attacker's control, user funds have already been compromised.

weiji-cryptonatty commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+There are various cases that we might want to export one of many private keys from a much securer but less convenient wallet, which is controlled with a seed or pass phrase.
+
+1.  We might dedicate one of many private keys for messaging purpose, and that private key is probably managed in a not-so-secure manner;
+2.  We might want to export one of many private keys from a hardware wallet, and split it with MPC technology so that a 3rd party service could help us identify potential frauds or known bad addresses, enforce 2FA, etc., meanwhile we can initiate transactions from a mobile device with much better UX and without carrying a hardware wallet.
```
Reviewing the motivation, however, it seems in this exact case we may have alternative request to R2 and R3, that is, each party $P_i$ in a MPC setting may generate its own ephemeral key pair $(r_i, R_i)$, and then Sender application may safely split `sk` safely into several pieces $sk_i$, then encapsulate each $sk_i$ to $R_i$. This effectively replaces the usual Distributed Key Generation procedure in a typical MPC protocol. 

To be fair, this may be added as an amendment to EIP-6051. I need to survey MPC protocols to see if the above suggestions work in their security models.

weiji-cryptonatty commented:
> I am going to mainly focus on the editorial review for now, so as we could unblock it for merging as `draft`, but I do have a few technical questions / feedback when conducting my technical peer review, and they will come later

Hi @xinbenlv any updates to the editorial review? Any outstanding issues? How can I be of assistance to help you help me? 

Also you are welcome to post any technical questions or feedback to the discussion thread. Many thanks.

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+## Motivation
+
+There are various cases that we might want to export one of many private keys from a much securer but less convenient wallet, which is controlled with a seed or pass phrase.
```
```suggestion
There are various cases in which we might want to export one of many private keys from a much more secure but less convenient wallet, which is controlled with a seed or passphrase.
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+## Abstract
+
+This EIP proposes a mechanism to encapsulate a private key so that it could be securely relocated to another application without providing the seed. This EIP combines `ECIES` (Elliptic Curve Integrated Encryption Scheme) and optional signature verification under various choices to ensure that the private key is encapsulated to a known or trusted party.
```
```suggestion
This EIP proposes a mechanism to encapsulate a private key so that it could be securely relocated to another application without providing the seed. This EIP combines `ECIES` (Elliptic Curve Integrated Encryption Scheme) and optional signature verification under various choices to ensure that the private key is encapsulated by a known or trusted party.
```

weiji-cryptonatty replied:
I was just suggest to use `for` here. It looks like using `by` will give people an impression that the known or trusted party executes the action of the encapsulation.

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+There are various cases that we might want to export one of many private keys from a much securer but less convenient wallet, which is controlled with a seed or pass phrase.
+
+1.  We might dedicate one of many private keys for messaging purpose, and that private key is probably managed in a not-so-secure manner;
```
```suggestion
1.  We might dedicate one of many private keys for messaging purposes, and that private key is probably managed in a not-so-secure manner;
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+### Sender and Recipient
+
+Sender is the party who has the private key to be encapsulated. Sender application is what Sender uses.
```
```suggestion
We hereby define:
- *Sender* as the party who holds in custody the private key to be encapsulated. *Sender Application* as the client-side application that said Sender uses.
```

weiji-cryptonatty replied:
Many thanks for the very careful, very detailed review, and revision suggestions! @xinbenlv

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+Sender is the party who has the private key to be encapsulated. Sender application is what Sender uses.
+
+Recipient is the party who accepts the encapsulated private key, unwrap and then use it. Recipient application is what Recipient uses.
```
```suggestion
- *Recipient* as the party who accepts the encapsulated private key, unwrap and then use it. *Recipient Application* as the client-side application that Recipient uses to receive the encapsulated private key.
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+### Core Algorithms
+
+The basic idea is to encapsulate the private key with ECIES. To ensure that the ephemeral public key to encapsulate the private key to is indeed generated from a trusted party and has not been tampered with, an option is provided to sign that ephemeral public key.
```
```suggestion
The basic idea is to encapsulate the private key with ECIES. To ensure that the ephemeral public key to encapsulate the private key is indeed generated from a trusted party and has not been tampered with, we also provided an option to sign that ephemeral public key in this protocol.
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+The basic idea is to encapsulate the private key with ECIES. To ensure that the ephemeral public key to encapsulate the private key to is indeed generated from a trusted party and has not been tampered with, an option is provided to sign that ephemeral public key.
+
+There should be a mandatory  `version`  parameter. This allows various kinds of Key Encapsulation Mechanisms to be adopted depending on the security considerations or preference. The list shall be short to minimize compatibility issues among different vendors.
```
```suggestion
There should be a mandatory  `version`  parameter. This allows various kinds of Key Encapsulation Mechanisms to be adopted depending on the security considerations or preferences. The list shall be short to minimize compatibility issues among different vendors.
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+There should be a mandatory  `version`  parameter. This allows various kinds of Key Encapsulation Mechanisms to be adopted depending on the security considerations or preference. The list shall be short to minimize compatibility issues among different vendors.
+
+In addition to `version` string, parameters, involved keys and functions include:
```
```suggestion
In addition to a `version` parameter, the following keys and functions are involved:
```

xinbenlv commented on EIPS/eip-6051.md:
````diff
@@ -0,0 +1,469 @@
+// expected return value: R
+```
+
+`signerPubKey`  is optional. If provided, it is assumed that the implementation has the corresponding private key and the implementation MUST sign the ephemeral public key (in the form of what is to be returned). The signature algorithm is determined by the curve part of the  `version`  parameter, that is, ECDSA for secp256k1, and Ed25519 for Curve25519. And in this situation, it should be the ca ...
````
```suggestion
`signerPubKey`  is optional. If provided, it is assumed that the implementation has the corresponding private key and the implementation MUST sign the ephemeral public key (in the form of what is to be returned). The signature algorithm is determined by the curve part of the  `version`  parameter, that is, ECDSA for secp256k1, and Ed25519 for Curve25519. And in this situation, it should be the case that the sender trusts  `signerPubKey`, no matter how this trust is maintained. If not, the next request WILL be rejected by Sender application. Also see **Security Considerations**.
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+`account`  is used to identify which private key to be encapsulated. With Ethereum it is an address. Also see **Encoding of data and messages**.
+
+If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format etc. SHALL fail the call and result in empty return and optional error logs.
```
```suggestion
If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format etc. SHALL fail the call and result in empty return and optional error logs.
```
```suggestion
If  `signerPubKey`  is provided or  `recipient`  contains `signature` data, the implementation MUST perform signature verification. Missing data or incorrect format MUST either fail the call or result in empty return and optional error logs.
```

I am not sure if "SHALL fail the call or return error with error logs" is logically sound here. Changing "MUST either fail the call, or return empty with error logs"

weiji-cryptonatty replied:
Good catch.

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+## Rationale
+
+A crucial difference of this [EIP-6051](./eip-6051.md) with [EIP-5630](./eip-5630.md) is that, as the purpose of key encapsulation is to transport private key securely, the public key from the key-recipient should be ephemeral, and mostly used only one-time. While in EIP-5630 settings, the public key of message recipient shall be stable for a while so that message senders can encrypt messages wit ...
```
```suggestion
A critical difference of this [EIP-6051](./eip-6051.md) with [EIP-5630](./eip-5630.md) is that, as the purpose of key encapsulation is to transport a private key securely, the public key from the key-recipient should be ephemeral, and mostly used only one-time. While in EIP-5630 settings, the public key of the message recipient shall be stable for a while so that message senders can encrypt messages without key discovery every time.
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+A crucial difference of this [EIP-6051](./eip-6051.md) with [EIP-5630](./eip-5630.md) is that, as the purpose of key encapsulation is to transport private key securely, the public key from the key-recipient should be ephemeral, and mostly used only one-time. While in EIP-5630 settings, the public key of message recipient shall be stable for a while so that message senders can encrypt messages wit ...
+
+There is security implication to this difference, including perfect forward secrecy. We aim to achieve perfect forward secrecy by generating ephemeral key pairs in both sides every time: 1) first the recipient shall generate an ephemeral key pair, retain the private key securely, and export the public key; 2) then the key sender securely wrap the private key in ECIES, with another ephemeral key p ...
```
```suggestion
There is security implication to this difference, including perfect forward secrecy. We aim to achieve perfect forward secrecy by generating ephemeral key pairs in both sides every time: 

1) first the recipient shall generate an ephemeral key pair, retain the private key securely, and export the public key;
2) then the key sender securely wrap the private key in ECIES, with another ephemeral key pair, then destroy the ephemeral key securely;
3) finally the recipient can unwrap the private key, then destroy its ephemeral key pair securely. After these steps, the cipher text in transport intercepted by a malicious 3rd party, is no longer decrypt-able.
```

xinbenlv commented on EIPS/eip-6051.md:
````diff
@@ -0,0 +1,469 @@
+//signerPubKey
+pub: '0x035a5ca16997f9b9ead9572c9bde36c5dab584b17bc965cdd7c2945c776e981b0b'
+```
+Then it proceeds to verify second layer signature, also as ECDSA over secp256k1:
````
```suggestion
Then it proceeds to verify the second layer signature, also as ECDSA over secp256k1:
```

xinbenlv commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+
+**Security Level**:
+
+1. We are not considering post-quantum security. If quantum computer becomes a materialized threat, the underlying cipher of Ethereum and other L1 chains would have been replaced, and this EIP will be outdated then (as the EC part of ECIES is also broken).
```
```suggestion
1. We are not considering post-quantum security. If the quantum computer becomes a materialized threat, the underlying cipher of Ethereum and other L1 chains would have been replaced, and this EIP will be outdated then (as the EC part of ECIES is also broken).
```

xinbenlv replied:
I think this is generally omittable because the whole Ethereum rely on ECDSA / secp256k1, if ECDSA is no longer safe due to quantum security, there are much bigger threat than the encapsulation described here.

weiji-cryptonatty replied:
I'd like to retain this part as part of the justifications for the current design.

xinbenlv reviewed (changes requested):
These are my main editorial review comments, many of them are due to the grammar issue. I totally understand because I speak English as second language too. Please consider run via a grammar checker. 

Another major issue is to keep it consistent for terminologies that you specifically defined. For example, when addressing "Recipient" or "Sender", please capitalize.

weiji-cryptonatty commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,469 @@
+In addition to `version` string, parameters, involved keys and functions include:
+
+1.  Sender private key  `sk`, to be encapsulated to recipient, and the corresponding address `account`.
+2.  Ephemeral recipient key pair  `(r, R)`  such that `R = [r]G`. `G` denotes the base point of the elliptic curve, and `[r]G` denotes the scalar multiplication. Optionally, `R` could be signed, and `signerPubKey` and `signature` are then provided for sender to verify if `R`  could be trusted or not.
```
Good catch. Fixed and a commit pushed.

weiji-cryptonatty commented:
> These are my main editorial review comments, many of them are due to the grammar issue. I totally understand grammar challenge myself because I speak English as second language. Please consider run via a grammar checker.
> 
> Another major issue is to keep it consistent for terminologies that you specifically defined. For example, when addressing "Recipient" or "Sender", please capitalize.

Hi @xinbenlv thank you again for the very careful and very detailed reviews. I can see why EIP editors are so overwhelmed. I really appreciate what you had done here.

I think I have fixed all issues. I also ran another round with grammarly.com and fixed almost all issues it suggests (with very few I did not agree with it). Please review again when you do find time.

xinbenlv commented:
From a technical perspective, I still think, if the EIP is defining a way to encapsulate and unwrap private key via `eth_` interface, that is considered an interface category, but we can address it later

xinbenlv reviewed (dismissed):
I think this is good to merge as a draft now. Defer to @Pandapip1 @SamWilsn to follow up the merging.

weiji-cryptonatty commented:
> I think this is good to merge as a draft now. Defer to @Pandapip1 @SamWilsn to follow up the merging.

Pandapip1 commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,474 @@
+discussions-to: https://ethereum-magicians.org/t/private-key-encapsulation-to-move-around-securely-without-entering-seed/11604
+status: Draft
+type: Standards Track
+category: ERC
```
```suggestion
category: Interface
```
This should be an Interface EIP, since it adds RPC endpoints.

Pandapip1 commented on EIPS/eip-6051.md:
````diff
@@ -0,0 +1,474 @@
+})
+```
+The return value should be `0x001d3f1ef827552ae1114027bd3ecf1f086ba0f9`. This matches the `account` parameter in **R2**.
+
+## Reference Implementation
+
+Contributions are welcome.
+
+The test vector generator is provided earlier in the **Test Cases** section.
````
```suggestion
```
This is not a reference implementation

Pandapip1 commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,474 @@
+
+## Security Considerations
+
+**Perfect Forward Secrecy**: PFS is achieved by using ephemeral key pairs on both sides.
```
Please use headings to indicate headings, not bold text.

Pandapip1 commented on EIPS/eip-6051.md:
```diff
@@ -0,0 +1,474 @@
+5.  *Recipient Application* decrypts  `cipher`  to get  `sk`.
+6.  *Recipient Application* derives the address corresponding to  `sk`  so that the user can confirm the correctness.
+
+With signature and verification, the signature to `R` by `singerPubKey` is appended to `R`. `signerPubKey` itself could have been already signed by `trustedPubKey`, and that signature is appended to `signerPubKey`. Note that the signature is applied to the byte array data instead of its string representation, which might lead to confusion and interoperability issues (such as hex or base64, lower  ...
```
Why not link to them?
```suggestion
With signature and verification, the signature to `R` by `singerPubKey` is appended to `R`. `signerPubKey` itself could have been already signed by `trustedPubKey`, and that signature is appended to `signerPubKey`. Note that the signature is applied to the byte array data instead of its string representation, which might lead to confusion and interoperability issues (such as hex or base64, lower case v.s. upper case, etc.). See [Requests](#requests) and [Test Cases](#test-cases) for further clarification and examples.
```

weiji-cryptonatty reviewed (commented):
suggestions accepted

Pandapip1 reviewed (approved):
LGTM for a draft
